client/core.go

```diff
@@ -0,0 +1,271 @@
+package client
+
+import (
+	"context"
+	"errors"
+	"net"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"github.com/gofiber/fiber/v3"
+	"github.com/gofiber/fiber/v3/addon/retry"
+	"github.com/valyala/fasthttp"
+)
+
+var boundary = "--FiberFormBoundary"
+
+// RequestHook is a function that receives Agent and Request,
+// it can change the data in Request and Agent.
+//
+// Called before a request is sent.
+type RequestHook func(*Client, *Request) error
+
+// ResponseHook is a function that receives Agent, Response and Request,
+// it can change the data is Response or deal with some effects.
+//
+// Called after a response has been received.
+type ResponseHook func(*Client, *Response, *Request) error
+
+// RetryConfig is an alias for config in the `addon/retry` package.
+type RetryConfig = retry.Config
+
+// addMissingPort will add the corresponding port number for host.
+func addMissingPort(addr string, isTLS bool) string { //revive:disable-line:flag-parameter // Accepting a bool param named isTLS if fine here
+	n := strings.Index(addr, ":")
+	if n >= 0 {
+		return addr
+	}
+	port := 80
+	if isTLS {
+		port = 443
+	}
+	return net.JoinHostPort(addr, strconv.Itoa(port))
+}
+
+// `core` stores middleware and plugin definitions,
+// and defines the execution process
+type core struct {
+	client *Client
+	req    *Request
+	ctx    context.Context //nolint:containedctx // It's needed to be stored in the core.
+}
+
+// getRetryConfig returns the retry configuration of the client.
+func (c *core) getRetryConfig() *RetryConfig {
+	c.client.mu.RLock()
+	defer c.client.mu.RUnlock()
+
+	cfg := c.client.RetryConfig()
+	if cfg == nil {
+		return nil
+	}
+
+	return &RetryConfig{
+		InitialInterval: cfg.InitialInterval,
+		MaxBackoffTime:  cfg.MaxBackoffTime,
+		Multiplier:      cfg.Multiplier,
+		MaxRetryCount:   cfg.MaxRetryCount,
+	}
+}
+
+// execFunc is the core function of the client.
+// It sends the request and receives the response.
+func (c *core) execFunc() (*Response, error) {
+	resp := AcquireResponse()
+	resp.setClient(c.client)
+	resp.setRequest(c.req)
+
+	// To avoid memory allocation reuse of data structures such as errch.
+	done := int32(0)
+	errCh, reqv := acquireErrChan(), fasthttp.AcquireRequest()
+	defer func() {
+		releaseErrChan(errCh)
+	}()
+
+	c.req.RawRequest.CopyTo(reqv)
+	cfg := c.getRetryConfig()
+
+	var err error
+	go func() {
+		respv := fasthttp.AcquireResponse()
+		if cfg != nil {
+			err = retry.NewExponentialBackoff(*cfg).Retry(func() error {
+				if c.req.maxRedirects > 0 && (string(reqv.Header.Method()) == fiber.MethodGet || string(reqv.Header.Method()) == fiber.MethodHead) {
+					return c.client.client.DoRedirects(reqv, respv, c.req.maxRedirects)
```

Comment:
that looks very strange, maybe we should rethink the name
```go
c.client.client
```

Reply:
follow up ticket

Reply:
> follow up ticket

I changed it to `fasthttp`